### ISSUE TITLE
The Inconvenience of Progress

### ISSUE THREAD

**Issue by hueniverse** [hapijs/hapi#1606]
A couple months ago I decided to switch hapi from the typical annual big splashy release with dozens of breaking changes to a more steady [semi-quarterly schedule of major releases](https://github.com/spumko/hapi/issues/1479). In other words, publish about 6-8 major releases a year, each with a very small number of breaking changes to made keeping up with the latest code base easier and safer.

The thinking behind this release scheme is to provide people who want to keep up smaller bites then can easily fit into their product development cycle, while others can skip multiple releases and catch up all at once, getting the “Big Release Experience”.

The general reaction has been positive, until people started getting behind.
### "I missing out on bug fixes"

There is a valid downside to this rapid release scheme which is that staying behind also means missing out on bug fixes. It is true that catching up with backwards compatible changes is mostly trivial. But in many ways, it’s the same problem with the big annual releases because once that big release is out, your cost of migration when a major bug is found on your older (and no longer supported) version is much higher and much riskier.

There are still people using hapi v0.x and v1.x with dozens of known bugs fixed in later version. We simply don't have the resources to back port those fixes.
### "Breaking changes are evil and must be avoided at all cost"

Many platforms go out of their way to avoid breaking changes. It makes sense for platforms people depend on, and hapi fits that criterion. But hapi is also relatively young, and is only now getting the wider adoption a framework requires to figure out its childhood mistakes.

Two of the last three major release of hapi (v4.0, v2.0) focused on validation. hapi uses a module called joi for validation of inputs and outputs. Validation is a core feature and a very important one for security and stability. The problem is, we got the joi API and some of its architecture very wrong. It was great two years ago when it was first created but we learned a lot using it and there was just no way to fix those issues without breaking it.

The next version of hapi (v5.0) is planned to be another joi upgrade to version v4.0. This should be the last big change to joi this year. I think we got it right this time with a clean API, a long list of features people have been asking for over a year, and more extensive testing. We are even offering [cash awards for validation bugs](https://github.com/spumko/joi/issues/281).

Breaking changes are never fun, but I refuse to get stuck with bad APIs and incorrect architecture just to avoid breaking changes. Whenever possible, we deprecate a feature and add a new one, and only remove the old one after enough time has passed. But this is not always possible.

Keeping bad APIs just not to introduce breaking changes has significant cost. It causes bad design to "leak" to other parts of the system, and to make future changes even more costly.
### "Changes are ok as long as I get a detailed change log"

The recent requests for a CHANGELOG caused some frustration given my general dismissive attitude towards it (I should have explained myself better and with more patience - hence this note).

It’s not that we don’t provide detailed list of all changes, it’s that we decided to provide them using the GitHub issues and milestones facilities. Some people find those too time consuming. Others find the issues text to be cryptic and lacking details.

I don’t have anything against adding a CHANGELOG (and we did accept a pull request adding an initial version of one). My argument is that those who want a better format should do the work. If you don’t understand a change, ask. That’s why we have issues and comments and an IRC channel and Twitter. The hapi team is probably the most accessible bunch of folks in the entire node community.

I would like to see some sort of automation that takes issues and milestones and generate release notes. It would be great if someone helped with that. But given our limited time, I much rather focus on writing code (as long as the change log data is available) and let others contribute by making sense of the issues data.

Also, every major release has a specially marked issue with all the breaking changes summed up. We’re not monsters.
### "Rapid releases means the product is just not ready for prime time"

Well, it depends. hapi is stable enough for Walmart to put a ~$15B business on it. Our team has to keep up with the major release just like everyone else, and it is as much pain for us as it is for everyone else. We are usually the first to find major issues. But if your schedule, attention span, or philosophy require a framework that never changes or changes very infrequently, then maybe hapi is not for you.

I think hapi is awesome, mostly because of everyone else working on it and the wonderful community that has evolved around it. I am also pretty proud of my own contribution. I am willing to accept that the decision I’ve made to do quick major releases is going to cost hapi some adoption. But this is not a right or wrong issue but a matter of priorities and resource management. Over a 5 years period you will spend more time upgrading annual releases than smaller, more frequent ones.
### It’s about resources

My main goal for this year is to bring new maintainers from other companies (or individuals). We added @nlf a few months ago as the first and we are trying to get other companies using hapi to join in. As the community around hapi grows and moves away from being Walmart-centric, so will its style and policies.

In other words, the way to accomplish the change you are looking for it to roll up your sleeves and help.
